odoo/model.go

```diff
@@ -0,0 +1,92 @@
+package odoo
+
+import "fmt"
+
+// SearchReadModel is used as "params" in requests to "dataset/search_read" endpoints.
+type SearchReadModel struct {
+	Model  string   `json:"model,omitempty"`
+	Domain []Filter `json:"domain,omitempty"`
+	Fields []string `json:"fields,omitempty"`
+	Limit  int      `json:"limit,omitempty"`
+	Offset int      `json:"offset,omitempty"`
+}
+
+// Filter to use in queries, usually in the format of
+// [predicate, operator, value], eg ["employee_id.user_id.id", "=", 123]
+type Filter interface{}
+
+// Method identifies the type of write operation.
+type Method string
+
+const (
+	// MethodWrite is used to update existing records.
+	MethodWrite Method = "write"
+	// MethodCreate is used to create new records.
+	MethodCreate Method = "create"
+	// MethodRead is used to read records.
+	MethodRead Method = "read"
+	// MethodDelete is used to delete existing records.
+	MethodDelete Method = "unlink"
+)
+
+// WriteModel is used as "params" in requests to "dataset/create" or "dataset/write" endpoints.
+type WriteModel struct {
+	Model  string `json:"model"`
+	Method Method `json:"method"`
+	// Args contains the record to create or update.
+	// If Method is MethodCreate, then the slice may contain a single entity without an ID parameter.
+	// Example:
+	//  Args[0] = {Name: "New Name"}
+	// If Method is MethodWrite, then the first item has to be an array of the numeric ID of the existing record.
+	// Example:
+	//  Args[0] = [221]
+	//  Args[1] = {Name: "Updated Name"}
+	Args []interface{} `json:"args"`
+	// KWArgs is an additional object required to be non-nil, otherwise the request simply fails.
+	// In most cases it's enough to set it to `map[string]interface{}{}`.
+	KWArgs map[string]interface{} `json:"kwargs"`
+}
```

Comment:
I'm not sure about this Model. It feels a bit very "free-form".
Alternative: Make dedicated `CreateModel`, `UpdateModel` and `DeleteModel` types since their usage in the `args` fields differ for each type of operation.
Or maybe this type is entirely package-internal so that public Method become simply like `CreateGenericModel(ctx, session, model string, id int, data interface{}`)?

Reply:
I'm fine with it as-is, but I'd make a comment in the form of: `Should always be created through New*Model.`

Then you could make the fields private. Making the entire model private would not be nice to use as soon as you want to pass it in another package.

Reply:
I would probably prefer to keep this type package-internal.

Reply:
I was able to make the generic model methods more developer friendly without requiring the WriteModel.
Still I left it publicly accessible so that it's possible to run `ExecuteQuery` if someone needs this.